newton/_src/sim/builder.py

```
@@ -4962,125 +4969,456 @@ def add_rod(
+            p1 = node_positions[v]
+            seg_vec = p1 - p0
+            seg_length = float(wp.length(seg_vec))
+            if seg_length <= min_segment_length:
                 raise ValueError(
-                    "add_rod: quaternion at index "
-                    f"{i} does not align capsule +Z with segment (positions[i+1] - positions[i]); "
-                    "quaternions must be world-space and constructed so that local +Z maps to the "
-                    "segment direction positions[i+1] - positions[i]."
+                    f"add_rod_graph: edge {e_idx} has a too-small length (length={seg_length:.3e}); "
+                    f"segment length must be > {min_segment_length:.1e}"
                 )
```

Ruff (0.14.14)

[warning] Avoid specifying long messages outside the exception class

(TRY003)